Some fixes for 3ds Max 2021

JenusL opened the pull request:
I found some issues so I will keep adding to this PR but wanted to open it asap so we could have an discussion as well.

Found issues
* Setting VRay threads need to be an int. **FIXED**
* Bunch of other VRay settings needs to be converted to correct type. There was also a copy/pasta error that I fixed. **FIXED**
* UNC fileout paths comes out with three slashes in beginning causing rendering failure.
Problem seems to be excessive use of `fileout = fileout.replace("\\", "\\\\")` at several places in the code. **FIXED**
* Duplicate option in kso_3dsmax.ini throws an error in Python 3. **FIXED**
* VRay isn't automatically started in silent mode under 3dsmaxbatch.exe.
This causes rendering to halt if any VRay pop up is opened and waiting for a response. Overwriting the placeholder file is one of them causing VRay to always halt indefinitely. But other cases could be a pop up on scene open telling that this was created in an older version, etc.
The fix is to call setVRaySilentMode(). To fix the issue with popups on scene load, this need to be called before the scene is opened, which means it will be called even if the scene doesn't use VRay at all. **FIXED**

JenusL committed:
vray num threads needs to be int

JenusL committed:
fix more vray settings to be correct type

h-schoenberger commented:
Hi



1. The int and type changes are probably required for the change to pymxs.

2. I remember that the UNC slashes had been required. Might not be required for pymxs as well.

3. An issue without setVRaySilentMode did not occur until now. RR has a popup dectection as well like handling this window:
RenderAppWindow_WIN= Render history settings ~ The path ~ 1 ~ Button ~ &No
So far there was no customer report about any other VRay dialogs stopping the render, but calling this function might be a good idea. But I do not know if VRay will then fail or continue.
As stated in the docs:  "Disables any V-Ray dialogues that may appear when loading a scene (i.e. notification for differences with older versions, etc.)." What happens if there is a wrong VRay version?
Needs to be tested. Perhaps RR has to compare the VRay version and exit in such a case.

PS: kso_3dsmax.ms is not used any more. 3dsmaxbatch is more stable than 3dsmaxcmd.exe

JenusL commented:
1. More type changes fixed now.
2. I will test to see how the UNC-paths work. But I have never used such a hack in my Python code so might have been a workaround to be able to execute a maxscript string in the default kso.
3. I updated the first post a bit and removed the part about kso_3dsmax.ms since I noticed it wasn't used any more :)
Yeah but how many users are using the 3dsmaxbatch?
Here's the window I get in interactive mode. 
![vray_raw_overwrite_placeholder](https://user-images.githubusercontent.com/23527584/81398062-cbac2f80-9128-11ea-9c30-9e27b45ff104.png)
Maybe it's better to add it to the global config with the rest of the window handlers instead of using setVRaySilentMode()

JenusL commented:
One more issue that I already fixed in the .zip file I sent you...
`configparser` in Python 3 is strict as default so it throws `DuplicateOptionError: While reading from 'C:\\RR_localdata\\temp\\A\\kso_3dsmax.ini'`
Cause of this is in fact that `ElementsFolder:	*ElementsFolder*` is duplicate on both line 13 and 31 in kso_3dsmax.ini.
I'll fix this again here.

JenusL committed:
remove duplicate ElementsFolder option

h-schoenberger commented:
>Yeah but how many users are using the 3dsmaxbatch?
It is the default for 3dsmax 2021 in RR now.
3dsmaxbatch 2019 had some issues and we did not switch, but the current tests look good.

>Here's the window I get in interactive mode.
I am not sure if VRay does not print any dialog if you are running from commandline.
It usually happens with rarely used plugins. The default 3dsmaxcmd.exe execution (without our script) does not allow to set any VRay option/function and therefore VRay had to take care that dialogs are not shown in commandline mode.

RR has the option to re-render existing frames and as far as I remember a 3dsmax-VRay company has used it. Note: 3dsmaxbatch has an option that disables all dialog windows no matter the plugin.

JenusL committed:
remove slashes conversion for python

JenusL commented:
> RR has the option to re-render existing frames and as far as I remember a 3dsmax-VRay company has used it. Note: 3dsmaxbatch has an option that disables all dialog windows no matter the plugin.

Ok so what do you think is the best approach then?

h-schoenberger commented:
>Ok so what do you think is the best approach then?
I assume it has no effect calling setVRaySilentMode().
But on the other side it has no disadvantage as well. Just add it, then it can be used inside the 3dsmax interactive UI for testing.

JenusL commented:
> I assume it has no effect calling setVRaySilentMode().
> But on the other side it has no disadvantage as well. Just add it, then it can be used inside the 3dsmax interactive UI for testing.

No no. `setVRaySilentMode()` has effect but I got the impression you were against it and wanted to add window handling instead. So that's why I asked.

JenusL commented:
Here's another one having sort of the same issue: https://forums.autodesk.com/t5/3ds-max-programming/batch-processing-was-blocked/td-p/8831336
My best guess is that Chaosgroup hasn't added "auto-silent" mode when running under 3dsmaxbatch.exe, but they did have it under 3dsmaxcmd.exe.
What else can explain this issue?

JenusL commented:
Or it could be a new behavior in newer versions of V-Ray Next. I don't know.
The only thing I do know is that without any changes it won't render anything if it needs to overwrite a file, being a placeholder or previous render.

h-schoenberger commented:
>My best guess is that Chaosgroup hasn't added "auto-silent" mode when running
>under 3dsmaxbatch.exe, but they did have it under 3dsmaxcmd.exe.

Yes, that sounds right

JenusL committed:
set vray to silent mode

JenusL commented:
Added `setVRaySilentMode()` properly now.
One problem now is exit code -130. I see that you have added an ignore parameter to rrCheckexitcode.exe. That ignore doesn't work for me so I guess I need an updated exe for that. Right?
Have you reported a bug that python exits with -130 to Autodesk? Or should I do it?

I will do some additional testing now. To see that everything else works as well.
BTW, have you had any time to look at some of the FIXMEs?

JenusL committed:
change logging type to set for vray silent mode

JenusL committed:
round printed gamma values for nicer logging

h-schoenberger commented:
>One problem now is exit code -130
Yes, I had a download link in the rrKnights Tavern thread. And a new release is about to be uploaded right now.

I have reported the issue via the public development forum as the ADN information stated that this forum is monitored by develpers. (I had multiple very bad experiences if I try to use the Autodesk support to get to the developers. Even if they completly broke the batch render. As you can see in rrHelp section 3dsmax/errors, versions stated at the bottom)

>BTW, have you had any time to look at some of the FIXMEs?
Not yet. But I remember at least for the VRay get element name that it returned None if you had never set any filename. But I have to test it. At least I do not have to take care about the pre-VRay-Next versions for 3dsmax 2021.

JenusL commented:
Done some testing and everything seems ok.
The only worry I have is for two of the FIXMEs:
https://github.com/BinaryAlchemy/RoyalRender2020_render_apps/blob/fc51e53b29ebe9c062f3363927613b0159ad3a69/scripts/kso_3dsmax_pymxs.py#L747-L750

That code should run `if arg.RendererMode == "GIPrePass"`
So my guess is that when doing GI prepass, it should disable framebuffer and all outputs, so I included raw as well.
So I would change it to this:
```python
        logMessageSET("VRay Raw Off")
        vray_settings.output_saveRawFile = False
        logMessageSET("VRay Seperate Render Channels Off")
        vray_settings.output_splitgbuffer = False
        logMessageSET("VRay Framebuffer Off")
        vray_settings.output_on = False
```
What do you think?

JenusL committed:
Revert "remove duplicate ElementsFolder option"

This reverts commit 9a996d9ea5aeecc8bfa0755d03c0fb9eb5cdcab5.

JenusL commented:
I saw that you had changed the kso_3dsmax.ini file yourself so I reverted the commit I did so they wont conflict.

h-schoenberger commented:
>That code should run if arg.RendererMode == "GIPrePass"

If VRay render mode is set to GI Prepass (only), then it does not matter what you have set as render output as VRay does not execute it. It just renders the GI map.

JenusL commented:
> > That code should run if arg.RendererMode == "GIPrePass"
> 
> If VRay render mode is set to GI Prepass (only), then it does not matter what you have set as render output as VRay does not execute it. It just renders the GI map.

Ok so we just leave it then?

h-schoenberger commented:
But you are right, the log messages seem to be wrong.
I have to do more tests to know why the output settings had been changed.

h-schoenberger commented:
All I know is that the GI prepass render was tested and the right files had been produced.

JenusL commented:
> All I know is that the GI prepass render was tested and the right files had been produced.

I will update to .30 on Monday and se what we have and I'll test the GI prepass things as well on Monday.
Have a nice weekend!

JenusL committed:
fix GIPrePass VFB logic

JenusL commented:
Updated to .30 and it works good with the ignore -130 error code now.
I also investigated the prePass stuff and that is working as I thought. It will save out a vrayRaw image if that is checked but it disables splitrgb if that is checked. That feels wrong to me and it should be strict and disable both.
Disabling VFB will disable both outputs, but that will also disable custom resolution if that is specified in VRayVFB instead of using Maxs resolution (Get resolution from MAX is unchecked) so that's not good.
I have committed a change of how I thought it should work. Let me know what you think of it and if that works for you as well.

h-schoenberger commented:
Hi

I have done some tests.
The default 3dsmax output is disabled by VRay if Irradieance Pre-render is enabled.
Same as it does in Maya.

But if VRay Framebuffer is enabled, then it still saves these files.
VRay Framebuffer was either enabled for seperate channels or multi-exr files (VRay Raw output)
So the simpliest solution was to disable the VFB.

>Disabling VFB will disable both outputs, but that will also disable custom resolution
Something that was not thought of as it was rarely used. Why change it there if you can change the resolution in the main window.

I have taken a look and your solution to keep VFB enabled and disable Raw and split channels works fine.

I am merging the changes. 

PS: And as stated in the 3dsmax forum "return code -130", I will switch back to the seperate log file for our print messages to prevent this error. This way RR can still react to an -130 error code and there is no false error message "There have been errors" which confuses render wranglers.
